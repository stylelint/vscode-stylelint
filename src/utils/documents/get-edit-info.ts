import type { TextDocument, TextEdit } from 'vscode-languageserver-textdocument';
import type { LintDiagnostics } from '../stylelint/index';
import type LSP from 'vscode-languageserver-protocol';

/**
 * Get the edit info for a diagnostic.
 * @param document The document to get edit info for.
 * @param diagnostic The diagnostic to get edit info for.
 */
export function getEditInfo(
	document: TextDocument,
	diagnostic: LSP.Diagnostic,
	lintResult:
		| (LintDiagnostics & {
				version: number;
		  })
		| undefined,
): { label: string; edit: TextEdit } | undefined {
	if (!lintResult || document.version !== lintResult.version) {
		return undefined;
	}

	const warning = lintResult.getWarning?.(diagnostic);

	if (!warning) {
		return undefined;
	}

	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
	// @ts-ignore -- (TS2353) `fix` property is available since v16.15.
	const edit = warning.fix;

	if (!edit) {
		return undefined;
	}

	return {
		label: `Fix this ${warning.rule} problem`,
		edit: {
			newText: edit.text,
			range: {
				start: document.positionAt(edit.range[0]),
				end: document.positionAt(edit.range[1]),
			},
		},
	};
}
