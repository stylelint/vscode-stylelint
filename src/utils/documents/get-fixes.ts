import { createTextEdits } from './create-text-edits';
import type stylelint from 'stylelint';
import type { TextDocument, TextEdit } from 'vscode-languageserver-textdocument';
import type { StylelintRunner, RunnerOptions } from '../stylelint/index';

/**
 * Runs Stylelint and returns fix text edits for the given document.
 * @param runner The Stylelint runner.
 * @param document The document to get fixes for.
 * @param linterOptions Linter options to use.
 * @param runnerOptions The runner options.
 */
export async function getFixes(
	runner: StylelintRunner,
	document: TextDocument,
	linterOptions: stylelint.LinterOptions = {},
	runnerOptions: RunnerOptions = {},
): Promise<TextEdit[]> {
	const result = await runner.lintDocument(
		document,
		{ ...linterOptions, fix: true },
		runnerOptions,
	);

	return typeof result.output === 'string' ? createTextEdits(document, result.output) : [];
}
