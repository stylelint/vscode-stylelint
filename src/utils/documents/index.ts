export * from './create-text-edits';
export * from './get-edit-info';
export * from './get-disable-type';
export * from './get-fixes';
export * from './get-workspace-folder';
