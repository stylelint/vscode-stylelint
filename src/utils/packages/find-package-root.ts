import fs from 'fs/promises';
import path from 'path';

/**
 * Walks up the file tree from the given path until it finds a directory
 * containing a file named `package.json`. Resolves to `undefined` if no such
 * directory is found.
 * @param startPath The path to start from.
 * @param rootFile The file to use to determine when the project root has been
 * reached. Defaults to `package.json`.
 */
export async function findPackageRoot(
	startPath: string,
	rootFile = 'package.json',
): Promise<string | undefined> {
	let currentDirectory = startPath;

	while (true) {
		const manifestPath = path.join(currentDirectory, rootFile);

		try {
			const stat = await fs.stat(manifestPath);

			if (stat.isFile()) {
				return currentDirectory;
			}

			const parent = path.dirname(currentDirectory);

			if (!path.relative(parent, currentDirectory)) {
				return undefined;
			}

			currentDirectory = parent;
		} catch (error) {
			if (
				(error as { code?: string }).code === 'ENOENT' ||
				(error as { code?: string }).code === 'ENOTDIR'
			) {
				const parent = path.dirname(currentDirectory);

				if (!path.relative(parent, currentDirectory)) {
					return undefined;
				}

				currentDirectory = parent;
			} else {
				throw error;
			}
		}
	}
}
