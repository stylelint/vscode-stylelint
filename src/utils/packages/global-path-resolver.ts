import os from 'os';
import path from 'path';
import type winston from 'winston';
import { runProcessFindLine } from '../processes';
import { PackageManager } from './types';

/**
 * Resolves the global `node_modules` path for different package managers.
 */
export class GlobalPathResolver {
	/**
	 * The logger to use for tracing resolution.
	 */
	#logger: winston.Logger | undefined;

	/**
	 * The cache of resolved paths.
	 */
	#cache: { [key in PackageManager]: string | undefined } = {
		yarn: undefined,
		npm: undefined,
		pnpm: undefined,
	};

	/**
	 * Whether or not the current platform is Windows.
	 */
	#isWindows: boolean;

	/**
	 * The resolvers by package manager.
	 */
	#resolvers: { [key in PackageManager]: () => Promise<string | undefined> } = {
		yarn: this.#yarn.bind(this),
		npm: this.#npm.bind(this),
		pnpm: this.#pnpm.bind(this),
	};

	/**
	 * Instantiates a new global path resolver.
	 * @param logger The logger to use for tracing resolution.
	 */
	constructor(logger?: winston.Logger) {
		this.#logger = logger;
		this.#isWindows = os.platform() === 'win32';
	}

	/**
	 * Resolves the global `node_modules` path for Yarn.
	 *
	 * Note: Only Yarn 1.x is supported. Yarn 2.x and higher have removed
	 * support for globally installed packages.
	 */
	async #yarn(): Promise<string | undefined> {
		const tryParseLog = (line: string): { type: string; data: string } | undefined => {
			try {
				return JSON.parse(line) as { type: string; data: string };
			} catch {
				return undefined;
			}
		};

		const yarnGlobalPath = await runProcessFindLine(
			'yarn',
			['global', 'dir', '--json'],
			this.#isWindows ? { shell: true } : undefined,
			(line) => {
				const log = tryParseLog(line);

				if (!log || log.type !== 'log' || !log.data) {
					return undefined;
				}

				const globalPath = path.join(log.data, 'node_modules');

				this.#logger?.debug('Yarn returned global node_modules path.', {
					path: globalPath,
				});

				return globalPath;
			},
		);

		if (!yarnGlobalPath) {
			this.#logger?.warn('"yarn global dir --json" did not return a path.');

			return undefined;
		}

		return yarnGlobalPath;
	}

	/**
	 * Resolves the global `node_modules` path for npm.
	 */
	async #npm(): Promise<string | undefined> {
		const npmGlobalPath = await runProcessFindLine(
			'npm',
			['config', 'get', 'prefix'],
			this.#isWindows ? { shell: true } : undefined,
			(line) => {
				const trimmed = line.trim();

				if (!trimmed) {
					return undefined;
				}

				const globalPath = this.#isWindows
					? path.join(trimmed, 'node_modules')
					: path.join(trimmed, 'lib/node_modules');

				this.#logger?.debug('npm returned global node_modules path.', {
					path: globalPath,
				});

				return globalPath;
			},
		);

		if (!npmGlobalPath) {
			this.#logger?.warn('"npm config get prefix" did not return a path.');

			return undefined;
		}

		return npmGlobalPath;
	}

	/**
	 * Resolves the global `node_modules` path for pnpm.
	 */
	async #pnpm(): Promise<string | undefined> {
		const pnpmGlobalPath = await runProcessFindLine(
			'pnpm',
			['root', '-g'],
			this.#isWindows ? { shell: true } : undefined,
			(line) => {
				const trimmed = line.trim();

				if (!trimmed) {
					return undefined;
				}

				this.#logger?.debug('pnpm returned global node_modules path.', {
					path: trimmed,
				});

				return trimmed;
			},
		);

		if (!pnpmGlobalPath) {
			this.#logger?.warn('"pnpm root -g" did not return a path.');

			return undefined;
		}

		return pnpmGlobalPath;
	}

	/**
	 * Attempts to resolve the global `node_modules` path for the given package
	 * manager.
	 *
	 * On a successful resolution, the method returns a promise that resolves to the
	 * package manager's global `node_modules` path. Paths are cached in the
	 * resolver on the first successful resolution.
	 *
	 * When a path cannot be resolved, the promise resolves to `undefined`.
	 *
	 * @example
	 * ```js
	 * const resolver = getGlobalPathResolver();
	 * const yarnGlobalPath = await resolver.resolve(
	 *   'yarn',
	 *   message => connection && connection.tracer.log(message)
	 * );
	 * ```
	 * @param packageManager The package manager to resolve the path for.
	 */
	async resolve(packageManager: PackageManager): Promise<string | undefined> {
		const cached = this.#cache[packageManager];

		if (cached) {
			return cached;
		}

		const resolver = this.#resolvers[packageManager];

		if (!resolver) {
			this.#logger?.warn('Unsupported package manager.', { packageManager });

			return undefined;
		}

		try {
			const globalPath = await resolver();

			if (globalPath) {
				this.#cache[packageManager] = globalPath;
			}

			return globalPath;
		} catch (error) {
			this.#logger?.warn('Failed to resolve global node_modules path.', {
				packageManager,
				error,
			});

			return undefined;
		}
	}
}
