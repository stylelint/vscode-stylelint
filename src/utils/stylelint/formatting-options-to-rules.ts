import type LSP from 'vscode-languageserver-protocol';

export type FormattingRules = {
	indentation?: [number | string];
	'no-missing-end-of-source-newline'?: true | null;
	'no-eol-whitespace'?: true | null;
};

/**
 * Converts the given formatting options to rules.
 * @param options The formatting options.
 * @returns The rules.
 */
export function formattingOptionsToRules({
	insertSpaces,
	tabSize,
	insertFinalNewline,
	trimTrailingWhitespace,
}: LSP.FormattingOptions): FormattingRules {
	// NOTE: There is no equivalent rule for trimFinalNewlines, so it is not respected.
	// TODO: Create respective rule upstream?

	const rules: FormattingRules = {
		indentation: [insertSpaces ? tabSize : 'tab'],
	};

	if (insertFinalNewline !== undefined) {
		rules['no-missing-end-of-source-newline'] = insertFinalNewline || null;
	}

	if (trimTrailingWhitespace !== undefined) {
		rules['no-eol-whitespace'] = trimTrailingWhitespace || null;
	}

	return rules;
}
