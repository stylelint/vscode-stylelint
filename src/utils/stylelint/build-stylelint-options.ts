import path from 'path';
import pathIsInside from 'path-is-inside';
import { URI } from 'vscode-uri';
import { findPackageRoot } from '../packages/index';
import type stylelint from 'stylelint';
import type { RunnerOptions } from './types';

/**
 * Given a document URI, base options, and extension options, builds a Stylelint
 * options object. Runner options supersede base options.
 */
export async function buildStylelintOptions(
	uri: string,
	workspaceFolder?: string,
	baseOptions: Partial<stylelint.LinterOptions> = {},
	{
		config,
		configFile,
		configBasedir,
		customSyntax,
		ignoreDisables,
		reportDescriptionlessDisables,
		reportNeedlessDisables,
		reportInvalidScopeDisables,
	}: RunnerOptions = {},
): Promise<Partial<stylelint.LinterOptions>> {
	const options = {
		...baseOptions,

		config: config ?? baseOptions.config,

		configFile: configFile
			? workspaceFolder
				? configFile.replace(/\$\{workspaceFolder\}/gu, workspaceFolder)
				: configFile
			: baseOptions.configFile,

		configBasedir: configBasedir
			? path.isAbsolute(configBasedir)
				? configBasedir
				: path.join(workspaceFolder ?? '', configBasedir)
			: baseOptions.configBasedir,

		customSyntax: customSyntax
			? workspaceFolder
				? customSyntax.replace(/\$\{workspaceFolder\}/gu, workspaceFolder)
				: customSyntax
			: baseOptions.customSyntax,

		ignoreDisables: ignoreDisables ?? baseOptions.ignoreDisables,

		reportDescriptionlessDisables:
			reportDescriptionlessDisables ?? baseOptions.reportDescriptionlessDisables,

		reportNeedlessDisables: reportNeedlessDisables ?? baseOptions.reportNeedlessDisables,

		reportInvalidScopeDisables:
			reportInvalidScopeDisables ?? baseOptions.reportInvalidScopeDisables,
	};

	const documentPath = URI.parse(uri).fsPath;

	if (documentPath) {
		if (workspaceFolder && pathIsInside(documentPath, workspaceFolder)) {
			options.ignorePath = path.join(workspaceFolder, '.stylelintignore');
		}

		if (options.ignorePath === undefined) {
			options.ignorePath = path.join(
				(await findPackageRoot(documentPath)) || path.parse(documentPath).root,
				'.stylelintignore',
			);
		}
	}

	return options;
}
