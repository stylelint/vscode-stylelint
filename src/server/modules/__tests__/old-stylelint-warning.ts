jest.mock('../../../utils/documents');
jest.mock('../../../utils/packages');
jest.mock('fs/promises');
jest.mock('path');

import fs from 'fs/promises';
import path from 'path';
import type LSP from 'vscode-languageserver-protocol';
import { TextDocument } from 'vscode-languageserver-textdocument';
import { getWorkspaceFolder } from '../../../utils/documents/index';
import { findPackageRoot } from '../../../utils/packages/index';
import { OldStylelintWarningModule } from '../old-stylelint-warning';
import { Stylelint } from '../../../utils/stylelint/index';

const mockedFS = fs as tests.mocks.FSPromisesModule;
const mockedPath = path as tests.mocks.PathModule;
const mockedGetWorkspaceFolder = getWorkspaceFolder as jest.MockedFunction<
	typeof getWorkspaceFolder
>;
const mockedFindPackageRoot = findPackageRoot as jest.MockedFunction<typeof findPackageRoot>;

let mockContext: ReturnType<typeof serverMocks.getContext>;
let mockLogger: ReturnType<typeof serverMocks.getLogger>;

describe('OldStylelintWarningModule', () => {
	beforeEach(() => {
		jest.clearAllMocks();
		mockedPath.__mockPlatform('posix');
		mockContext = serverMocks.getContext();
		mockLogger = serverMocks.getLogger();
		mockContext.__options.validate = [];
	});

	test('should be constructable', () => {
		expect(() => new OldStylelintWarningModule({ context: mockContext.__typed() })).not.toThrow();
	});

	test('onDidRegisterHandlers should register an onDidOpen handler', () => {
		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
		});

		module.onDidRegisterHandlers();

		expect(mockContext.documents.onDidOpen).toHaveBeenCalledTimes(1);
		expect(mockContext.documents.onDidOpen).toHaveBeenCalledWith(expect.any(Function));
	});

	test('if document is not part of a workspace, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue(undefined);
		mockContext.__options.validate = ['bar'];

		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.resolveStylelint).not.toHaveBeenCalled();
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Document not part of a workspace, ignoring',
			{ uri: 'foo' },
		);
	});

	test('if document has already been checked, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockContext.__options.validate = ['bar'];

		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const document = TextDocument.create('foo', 'bar', 1, 'a {}');
		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({ document });
		await handler({ document });

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Document has already been checked, ignoring',
			{ uri: 'foo' },
		);
	});

	test('if Stylelint package root cannot be determined, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue(undefined);
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {} as unknown as Stylelint,
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.__options.validate = ['bar'];

		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Stylelint package root not found', {
			uri: 'foo',
		});
	});

	test('if Stylelint package root is not in workspace, should not warn', async () => {
		const error = new Error('foo');

		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/usr/local/lib/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {} as unknown as Stylelint,
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.__options.validate = ['bar'];
		mockedFS.readFile.mockRejectedValue(error);

		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Stylelint package root is not inside the workspace',
			{ uri: 'foo' },
		);
	});

	test('if Stylelint package manifest cannot be read, should not warn', async () => {
		const error = new Error('foo');

		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {} as unknown as Stylelint,
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.__options.validate = ['bar'];
		mockedFS.readFile.mockRejectedValue(error);

		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Stylelint package manifest could not be read',
			{
				uri: 'foo',
				manifestPath: '/path/node_modules/stylelint/package.json',
				error,
			},
		);
	});

	test('if Stylelint package manifest is malformed, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {} as unknown as Stylelint,
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.__options.validate = ['bar'];
		mockedFS.readFile.mockResolvedValue('{');

		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Stylelint package manifest could not be read',
			{
				uri: 'foo',
				manifestPath: '/path/node_modules/stylelint/package.json',
				error: expect.any(Error),
			},
		);
	});

	test('if Stylelint package manifest does not contain a version, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {} as unknown as Stylelint,
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.__options.validate = ['bar'];
		mockedFS.readFile.mockResolvedValue('{}');

		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
	});

	test('if Stylelint version cannot be parsed, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {} as unknown as Stylelint,
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.__options.validate = ['bar'];
		mockedFS.readFile.mockResolvedValue('{"version": "foo"}');

		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Stylelint version could not be parsed', {
			uri: 'foo',
			version: 'foo',
			error: expect.any(Error),
		});
	});

	test('if Stylelint version is 14.x or greater, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {} as unknown as Stylelint,
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.__options.validate = ['bar'];
		mockedFS.readFile.mockResolvedValue('{"version": "14.0.0"}');

		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
	});

	test('if Stylelint version is 14.x with a label, should not warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {} as unknown as Stylelint,
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.__options.validate = ['bar'];
		mockedFS.readFile.mockResolvedValue('{"version": "14.0.0-sdk"}');

		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).not.toHaveBeenCalled();
	});

	test('without openDocument support, if Stylelint version is less than 14.x, should warn and provide link to migration guide', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {} as unknown as Stylelint,
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.__options.validate = ['bar'];
		mockedFS.readFile.mockResolvedValue('{"version": "13.0.0"}');

		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage.mock.calls[0]).toMatchSnapshot();
		expect(mockContext.connection.window.showDocument).not.toHaveBeenCalled();
	});

	test("with openDocument support, if Stylelint version is less than 14.x and user doesn't click button, should warn but not open URL", async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {} as unknown as Stylelint,
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.__options.validate = ['bar'];
		mockContext.connection.window.showWarningMessage.mockResolvedValue(undefined);
		mockedFS.readFile.mockResolvedValue('{"version": "13.0.0"}');

		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				window: {
					showDocument: { support: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showDocument).not.toHaveBeenCalled();
	});

	test('with openDocument support, if Stylelint version is less than 14.x and user clicks button, should warn and open URL', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {} as unknown as Stylelint,
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.__options.validate = ['bar'];
		mockContext.connection.window.showWarningMessage.mockResolvedValue({
			title: 'Open migration guide',
		});
		mockContext.connection.window.showDocument.mockResolvedValue({
			success: true,
		});
		mockedFS.readFile.mockResolvedValue('{"version": "13.0.0"}');

		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				window: {
					showDocument: { support: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showDocument).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showDocument.mock.calls[0]).toMatchSnapshot();
		expect(mockLogger.warn).not.toHaveBeenCalledWith('Failed to open migration guide');
	});

	test('with openDocument support, if Stylelint version is less than 14.x and user clicks button, but fails to open URL, should warn and log', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {} as unknown as Stylelint,
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.__options.validate = ['bar'];
		mockContext.connection.window.showWarningMessage.mockResolvedValue({
			title: 'Open migration guide',
		});
		mockContext.connection.window.showDocument.mockResolvedValue({
			success: false,
		});
		mockedFS.readFile.mockResolvedValue('{"version": "13.0.0"}');

		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				window: {
					showDocument: { support: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showDocument).toHaveBeenCalledTimes(1);
		expect(mockLogger.warn).toHaveBeenCalledWith('Failed to open migration guide');
	});

	test('if document language ID is not in options, should warn', async () => {
		mockedGetWorkspaceFolder.mockResolvedValue('/path');
		mockedFindPackageRoot.mockResolvedValue('/path/node_modules/stylelint');
		mockContext.resolveStylelint.mockResolvedValue({
			stylelint: {} as unknown as Stylelint,
			resolvedPath: '/path/node_modules/stylelint',
		});
		mockContext.__options.validate = ['baz'];

		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const handler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await handler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.resolveStylelint).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showWarningMessage).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.window.showDocument).not.toHaveBeenCalled();
	});

	it('should be disposable', () => {
		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		expect(module).toHaveProperty('dispose');
		expect(module.dispose).toBeInstanceOf(Function);
	});

	it('should dispose all handler registrations when disposed', () => {
		const module = new OldStylelintWarningModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();
		module.dispose();

		const disposable = mockContext.documents.onDidOpen.mock.results[0].value;

		expect(disposable.dispose).toHaveBeenCalledTimes(1);
	});
});
