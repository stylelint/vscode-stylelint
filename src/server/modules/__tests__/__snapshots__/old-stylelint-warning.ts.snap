// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OldStylelintWarningModule with openDocument support, if Stylelint version is less than 14.x and user clicks button, should warn and open URL 1`] = `
[
  {
    "external": true,
    "uri": "https://github.com/stylelint/vscode-stylelint#migrating-from-vscode-stylelint-0xstylelint-13x",
  },
]
`;

exports[`OldStylelintWarningModule without openDocument support, if Stylelint version is less than 14.x, should warn and provide link to migration guide 1`] = `
[
  "Stylelint version 13.0.0 is no longer supported. While it may continue to work for a while, you may encounter unexpected behavior. Please upgrade to version 14.0.0 or newer. See the migration guide for more information.",
]
`;
